### ISSUE TITLE
[java] Make SystemPrintln configurable

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#2197]
This rule could do with added granularity as to whether to include only System.err calls or both. It could also be modified to allow exclusion of these calls in any main() body.

https://stackoverflow.com/a/2750562/3924108

_Originally posted by @Fernal73 in https://github.com/pmd/pmd/issues/1602#issuecomment-570887797_

**Comment by Fernal73** [comment]
I imagine that this can be done by using an Xpath violation suppression as well. Checkstyle provides a command-line option that displays the AST tree for a single file. I know this can be done through the Rule Designer. Is there a possibility that PMD will allow for the AST tree to be output via the CLI? Better still, output the AST as an xml doc?

**Comment by oowekyala** [comment]
How would it be easier than to use the designer to inspect the tree? The designer has an XPath editor that uses PMD's XPath engine. The specific way we map Java values to XPath values is not standard and your XPath expression may not actually work the same way if you worked on an XML document.

Besides the designer is built specifically for this use case. If you have suggestions or complaints about its usability then open tickets there. Is the problem with launching the designer quickly? Would letting the designer be launched with a file name as a CLI parameter help? Is it that you prefer not to use the designer at all? (maybe you don't have javafx idk)

We could add an "export to XML" option in the designer. We could even let this XML document be dumpable from the designer CLI, without starting the GUI. But I don't think the value added to adding this to the PMD CLI would justify the added complexity.

**Comment by Fernal73** [comment]
No, I don't use to intend to use the UI, not from my phone.
Is it possible to use in non-UI mode?
Additionally, I find just four issues on the PMD Designer repository. This
makes me suspect that the tool is either not used as much as expected or
it's a brilliant piece of engineering that meets all the user's needs. The
two need not go hand in hand.

**Comment by oowekyala** [comment]
> No, I don't use to intend to use the UI, not from my phone.

So you want to read XML dumps on a phone screen... Not sure how relatable a use case this is.

> Additionally, I find just four issues on the PMD Designer repository. This makes me suspect that the tool is either not used as much as expected or it's a brilliant piece of engineering that meets all the user's needs. The two need not go hand in hand.

Your point? Since this discussion is off-topic, please don't comment about that anymore on this thread (don't email reply). Post replies on pmd/pmd-designer#36

**Comment by Fernal73** [comment]
Why does anything have to be relatable? Would you prefer I say command-line, instead? How does it matter? There has to be a simpler way of doing it.

**Comment by Fernal73** [comment]
Another easier way to suppress this rule is to include the method name in the violation message. That way, the error can be suppressed if the method is main().

**Comment by adangel** [comment]
> I imagine that this can be done by using an Xpath violation suppression as well.

Possibly. Usually we add custom properties to the rules, so that it is easier to enable/disable this enhancement, e.g.

```xml
<rule ref="category/java/bestpractices.xml/SystemPrintln">
    <properties>
        <property name="allowInMain" value="true" />
        <property name="allowSystemErr" value="true" />
    </properties>
</rule>
```

That's the topic of this issue: enhancing SystemPrintln.... 

Additionally, this would allow us to ship default behavior and even change it, if we think, it makes sense. Violation suppression (in whatever way, see https://pmd.github.io/latest/pmd_userdocs_suppressing_warnings.html) can be used in the mean time as a workaround.

**Comment by Fernal73** [comment]
Agreed. Violation suppression is generally a workaround or a customization
of a rule that users can't be bothered with to request a change for.
Speaking of which, Checkstyle allows users to customize violation messages,
if needed.

**Comment by Fernal73** [comment]
A workaround is to use the Logging facility of the JDK and set :

`java.util.logging.SimpleFormatter.format=%5$s%n`

in logging.properties.

This way, you have the logger and the output still looks like System.println output.

**Comment by Fernal73** [comment]
For now, I've set up the rule as: 
```xml
  <rule ref="category/java/bestpractices.xml/SystemPrintln">
    <properties>
      <property name="violationSuppressXPath" 
        value="./ancestor::MethodDeclaration[contains(@Name, 'main')]"/>
    </properties>
</rule>
```

**Comment by Fernal73** [comment]
I think we can safely assume that any class that has a main method should be exempt from this rule irrespective of whether the SystemPrintln calls are within the main method or other methods in that class. 
To that effect,
```xml
  <rule ref="category/java/bestpractices.xml/SystemPrintln">
    <properties>
      <property name="violationSuppressXPath" 
        value="//MethodDeclaration[contains(@Name, 'main')]"/>
    </properties>
</rule>

```

**Comment by Fernal73** [comment]
Or do we need more granularity so that only static methods in a class containing `main` are exempt?
